client-encryption/src/main/java/io/confluent/kafka/schemaregistry/encryption/FieldEncryptionExecutor.java

```diff
@@ -241,155 +175,235 @@ private static Object toObject(FieldContext fieldCtx, byte[] bytes) {
-            cryptor = getCryptor(ctx);
-            dek = getDekForEncrypt(kekId, cryptor.getDekFormat());
-            break;
-          case READ:
-            if (header == null) {
-              // Not encrypted
-              skip = true;
-              return;
-            }
-            setStateFromHeader(header.value());
-            break;
-          default:
-            throw new IllegalArgumentException("Unsupported rule mode " + ctx.ruleMode());
-        }
-      } catch (GeneralSecurityException | RuntimeException e) {
+        cryptor = getCryptor(ctx);
+        kekName = getKekName(ctx);
+        kek = getKek(ctx, kekName);
+        dek = getDek(ctx, kekName, kek);
+      } catch (GeneralSecurityException e) {
         throw new RuleException(e);
       }
     }
 
-    protected String getKekId(RuleContext ctx) {
-      String keyId = ctx.getParameter(ENCRYPT_KMS_KEY_ID);
-      if (keyId == null) {
-        keyId = defaultKeyId;
+    protected String getKekName(RuleContext ctx) throws RuleException {
+      String name = ctx.getParameter(ENCRYPT_KEK_NAME);
+      if (name == null) {
+        throw new RuleException("No kek name found");
       }
-      if (keyId == null) {
-        throw new IllegalArgumentException("No key id found");
+      int length = name.length();
+      if (length == 0) {
+        throw new RuleException("Empty kek name");
       }
-      return getKeyUrlPrefix(ctx) + keyId;
+      char first = name.charAt(0);
+      if (!(Character.isLetter(first) || first == '_')) {
+        throw new RuleException("Illegal initial character in kek name: " + name);
+      }
+      for (int i = 1; i < length; i++) {
+        char c = name.charAt(i);
+        if (!(Character.isLetterOrDigit(c) || c == '_' || c == '-')) {
+          throw new RuleException("Illegal character in kek name: " + name);
+        }
+      }
+      return name;
     }
 
-    protected Dek getDekForEncrypt(String kekId, DekFormat dekFormat) {
-      EncryptKey key = new EncryptKey(kekId, dekFormat);
-      try {
-        return dekEncryptCache.get(key);
-      } catch (ExecutionException e) {
-        throw new RuntimeException(e);
+    protected KekInfo getKek(RuleContext ctx, String kekName) throws RuleException {
+      KekId kekId = new KekId(kekName);
+      String kmsType = ctx.getParameter(ENCRYPT_KMS_TYPE);
+      String kmsKeyId = ctx.getParameter(ENCRYPT_KMS_KEY_ID);
+
+      KekInfo kek = retrieveKekFromRegistry(ctx, kekId);
+      if (kek == null) {
+        if (ctx.ruleMode() == RuleMode.READ) {
+          throw new RuleException("No kek found for " + kekName + " during consume");
+        }
+        if (kmsType == null) {
+          throw new RuleException("No kms type found for " + kekName + " during produce");
+        }
+        if (kmsKeyId == null) {
+          throw new RuleException("No kms key id found for " + kekName + " during produce");
+        }
+        kek = new KekInfo(kmsType, kmsKeyId, false);
```

Comment:
Sharing a KEK means that they've given SR the proper permissions/policies to access their KMS.  If the KEK is shared, then SR can decrypt the DEK; otherwise only the client code can decrypt the DEK.